Block Styles: Fix block style variation selector generation

aaronrobertshaw opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Updates the approach to generating selectors for block style variations so that they aren't broken if a block's selector happens to be an element tag such as the paragraph block.

## Why?

With the current approach just pretending `is-style-whatever` to the block's selector we end up with a broken selector for the paragraph block. As block style variations are being extended to support section styling it is highly likely theme authors and users might want to have some different styles within paragraphs on a per section basis.

## How?

- Creates a new function to add the block style variation's class to the block's current selector.
- Adds some unit tests to cover most use cases 

## Testing Instructions

1. Check `npm run test:unit:php:base -- --filter WP_Theme_JSON_Gutenberg_Test` is all green
2. Confirm block styles in the editor still work

aaronrobertshaw committed:
Block Styles: Fix block style variation selector generation

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3869,4 +3869,41 @@ function ( $carry, $item ) {
+	 *
+	 * @return string Block selector with block style variation selector added to it.
+	 */
+	public static function get_block_style_variation_selector( $variation_name, $block_selector ) {
```
Does this need to be a `public` function? It looks like most of the similar methods (`append_to_selector`, `prepend_to_selector`, etc) are `protected`.

aaronrobertshaw replied:
I figured this one warranted some tests given the regex in it and I wanted the ability to generate the variation selector in a hook if we end up going down the per-block instance variations styles path.

Do you think making it generic e.g. `inject_class` or something more appropriate, would be better? Then do the extra leg work to make the variation selectors where they are needed.

andrewserong replied:
> Do you think making it generic e.g. inject_class or something more appropriate, would be better? 

If making the method public, then yes, I think it'd be better to try to make it a bit more generic if we can. We're already a bit boxed in with the complexity of the Theme JSON class, but at least if the public API doesn't grow too much, there's the possibility of potentially simplifying things further down the track.

If going for `protected`, it'd still be great to include tests to cover the regex, but I'd probably lean toward optimising for a reduced public API surface area over the ease of testing. Or to put it another way, if we can do effectively the same tests, but moving "up" a couple of levels so we're testing the final output of the Theme JSON class rather than this particular method?

andrewserong replied:
Oh, one other thing in case it helps: if it's only set to `public` so that we can test the method, then we might be able to have it be `protected` and then use `ReflectionClass` and set the method to be accessible for your tests. I haven't tried this before, but found the following while looking through tests in `wordpress-develop`: https://github.com/WordPress/wordpress-develop/blob/2a6b527f73260925729beb7860a4a2a01513f527/tests/phpunit/tests/admin/wpSiteHealth.php#L44-L46

And: https://github.com/WordPress/wordpress-develop/blob/4ee594761f16e7aa9314fbf3088c3ad0c9ea3681/tests/phpunit/tests/hooks/removeAllFilters.php#L57-L63

aaronrobertshaw replied:
You make some good points @andrewserong 👍 

I get the suspicion that we will need to generate a block style variation selector outside of theme.json so the public option might be required. 🤔

I'm a bit short on good names for a more generic function. Do you have any ideas? 

- `add_class_to_ancestor`
- `add_class_to_selector_ancestors`
- `inject_class_in_selector`

In terms of related functions, `scope_selector` is public so I guess that's a precedent.

andrewserong replied:
Oh, naming it tough! 😄

I don't have any real preferences, but just throwing a few more into the ring:

* `add_class_to_outer_selector`
* `add_class_to_block_selector` (maybe too specific)
* `add_class_to_wrapper_selector`

Overall, I think I like the `add_class` prefix 👍

ChatGPT gave some similarly repetitive suggestions 🤔

1. `addWrapperClassToSelector`
2. `addClassToWrapperSelector`
3. `appendToWrapperSelector`
4. `updateWrapperSelectorClass`
5. `setWrapperClassInSelector`
6. `modifyWrapperSelectorWithClass`
7. `applyClassToWrapperSelector`
8. `prependClassToWrapperSelector`
9. `insertClassInWrapperSelector`
10. `setAncestorClassInSelector`
11. `addAncestorClassToSelector`
12. `updateAncestorSelectorClass`
13. `applyClassToAncestorSelector`
14. `prependClassToAncestorSelector`
15. `insertClassInAncestorSelector`

aaronrobertshaw replied:
I think the catch with `add_class_to_block_selector` is that the block's selector could be a compound and comma separated selector. It's why I tried to target the fact it is the ancestor that we're really trying to add the class to.

Now a few more minutes have passed. I'm going to sidestep the public API side of it. The new plan is:

1. Switch `get_block_style_variation_selector` to `protected`
2. Update the tests to call it via reflection class
3. When/if a hook does need that functionality, `get_block_style_variation_selector` can be refactored to have a generic public function do the heavy lifting which just gets used internally in `get_block_style_variation_selector`.

I've done the first to parts of this in https://github.com/WordPress/gutenberg/pull/58051/commits/1032daf873b826eb77529158c49143f30d938007. 

There's still the JS side of it which I'll get to shortly but I think the PHP side should be close now. Thanks for the help.

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3869,4 +3869,41 @@ function ( $carry, $item ) {
+						return $matches[0];
+					}
+
+					$class_added = true;
```
Does this always result in replacement happening in the first round of calling `preg_replace_callback`? Just wondering if this is much different to passing a `$limit` param of `1` to `preg_replace_callback` 🤔

Apologies if I'm missing nuance here!

aaronrobertshaw replied:
No, it was me missing things sorry. Just how it came out after a few failed attempts. 

Switching this to use `limit` cleans things up nicely.

andrewserong replied:
Nice. For what it's worth, I didn't know about the `limit` param until I started reviewing this and wondered how the `preg_replace_callback` function works and found it in the docs 😄

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3869,4 +3869,41 @@ function ( $carry, $item ) {
+		foreach ( $selector_parts as $part ) {
+			$class_added = false;
+			$result[]    = preg_replace_callback(
+				'/((?::is|not|has|where\([^)]+\))?\s*)([^\s:]+)/',
```
I'm having a little trouble wrapping my head around this regex, I'm sorry! Why do we need to name the valid pseudo-classes to match against?

From looking at the tests, it sounds like we always want the variation class name to be attached to the outer wrapper / the very first part of the selector, directly preceding any `:` character or whitespace?

Also, if it's being updated in PHP here, would we need a corresponding change in [the JS equivalent](https://github.com/WordPress/gutenberg/blob/e103afb02dc3c814cb95f37a85bc5117cc8458a1/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L1074)?

aaronrobertshaw replied:
> I'm having a little trouble wrapping my head around this regex, I'm sorry! 

So do I and I wrote it! 😅 

> Why do we need to name the valid pseudo-classes to match against?

I was thinking it might be necessary to restrict the potential pseudo functions that get matched but this is probably overkill. Without them works, passes the tests, and is likely more flexible. I've cleaned up the regex in [478ed9e](https://github.com/WordPress/gutenberg/pull/58051/commits/478ed9e4bc3c6a52d305b23c1d95d695a70b4531), thanks!

> Also, if it's being updated in PHP here, would we need a corresponding change in [the JS equivalent](https://github.com/WordPress/gutenberg/blob/e103afb02dc3c814cb95f37a85bc5117cc8458a1/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L1074)?

Tunnel vision on my part. I often get caught out with the split between PHP and JS.

andrewserong reviewed (commented):
Thanks for the ping @aaronrobertshaw! Just to make sure I'm following — the problem with the current approach on `trunk` is that anything that uses an element selector will be broken because it currently prepends the variation classname instead of appending, so we wind up getting `.is-style-customp` instead of `p.is-style-custom`.

Attempting to ensure that we're always adding the classname after the wrapping selector sounds like a good idea to me 👍. It's a bit tricky to predict, but I like the way you're honing in on what would be the valid characters before the bit we wish to inject.

Just left a few questions — my regex is a little rusty, so I copy / pasted it into https://regexr.com/ to play with some sample values, and I wasn't sure how much of the `is|not|has|where` we needed?

Also, it looks like we'll also need to update the JS logic for consistency. While manually testing, I noticed that the paragraph styles are output correctly for me in the post editor (which uses this PHP logic) but not in the site editor.

### Post editor

The selector is correct:

![image](https://github.com/WordPress/gutenberg/assets/14988353/064cb003-9fa0-4832-8531-175ba6443fc3)

### Site editor

This block is set to my `is-style-custom` style, but doesn't get the color:

![image](https://github.com/WordPress/gutenberg/assets/14988353/3d3bcfb3-06a5-4d4e-ba29-153fd8a06925)

Here's the output in the site editor:

<img width="266" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/f45739e9-2009-461a-8123-0a2ef81299f7">

To reproduce:

* I added the following at the end of the paragraph block's `block.json`:

```
	"styles": [
		{ "name": "fill", "label": "Fill", "isDefault": true },
		{ "name": "outline", "label": "Outline" },
		{ "name": "custom", "label": "Custom" }
	]
```

* Opened up the site editor, went to the Paragraph block in global styles, edited the Custom block style and gave it a color.
* Added a paragraph block and selected the Custom style.

aaronrobertshaw committed:
Clean up preg replace callback using limit

aaronrobertshaw committed:
Simplify regex

aaronrobertshaw committed:
Switch get_block_style_variation_selector to be protected

aaronrobertshaw committed:
Fix fat fingers toggling of hunk for commit

aaronrobertshaw committed:
Add JS version of function to generate variation selector

aaronrobertshaw commented:
I've quickly updated the JS side of block style variation selectors but am out of time today. I'll give this a proper test tomorrow before flagging it ready for a final review.

aaronrobertshaw committed:
Fix expected variation selector in js unit test

aaronrobertshaw commented:
I think this is ready for a final review. I fixed the expected selector in the JS unit tests I missed. 

The rest of the earlier unit test failures seemed unrelated but I've rebased this in case.

andrewserong commented:
Thanks for the update, Aaron! I likely won't get time to take this for another spin today, but if no-one beats me to it, I'll give this another review on Monday 🙂

andrewserong reviewed (approved):
Just managed to get this review in before wrapping up for the week. This is testing great for me now!

✅ Adding additional styles to the paragraph block as in my earlier review is now working nicely
✅ Paragraph and Button outline style updates are working well in the site editor, post editor, and on the site front end
✅ Regex pattern looks good to me, thanks for adding in the good test coverage for both the JS and PHP approaches!

<img width="911" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/7a2f8cc5-c336-4a98-89a1-7f8c069eb65b">

While it's good to avoid complex regex patterns where we can, in this instance I think parsing the CSS selector with these regex patterns is nicely justified:

* The custom selector is only set by blocks, so is unlikely to be super complex selectors and is relatively controlled
* Due to the above, we're not parsing user entered data
* We're not attempting to parse all of the CSS spec or anything like that

So, all up I think this is a good approach to go with — it resolves the current bug while making the feature more robust and adding some good test coverage.

LGTM! ✨

aaronrobertshaw committed:
Add test for wrapping :where

aaronrobertshaw commented:
Backport is available in https://github.com/WordPress/wordpress-develop/pull/6053

dmsnell commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3880,4 +3880,37 @@ function ( $carry, $item ) {
+
+		foreach ( $selector_parts as $part ) {
+			$result[] = preg_replace_callback(
+				'/((?::\([^)]+\))?\s*)([^\s:]+)/',
```
@aaronrobertshaw a PCRE pattern like this warrants an explanatory comment with an example showing an expected input and transformation.

dmsnell replied:
@aaronrobertshaw while trying to understand the purpose of this PCRE and coming up empty, I found this commit eacbd8a7adb7bf30fbe846a985d11d82cd360cf7 which claims to simplify the pattern, but I wonder if it actually breaks the pattern entirely?

is there any CSS we would want to match which looks like `:(…)`?
before the change there are recognizable segments like `:where(…)`, but after the commit I’m having trouble thinking up any valid CSS that would be detected with the pattern.

dmsnell replied:
it actually looks like the original PCRE before the “simplification” was also broken, but without knowing what it was meant to match, it’s hard to say.

I do not think it was originally meant to capture these segments:
 - `:is`
 - `not`
 - `has`
 - `where(…)`

And I do not think it is intended to match these segments now:
 - `:(.wp-block-quote)`
 - `:(p)`

And I think that because these are invalid CSS snippets.

The test with this string: `:where(.outer .inner:first-child)` produces an empty `$match[1]` and a `$match[2]` of `where(.outer` — is that what you wanted and expected? as it stands, I think we should only ever expect `$match[1]` to be empty, given that it would require invalid CSS to detect.

dmsnell commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3880,4 +3880,37 @@ function ( $carry, $item ) {
+				'/((?::\([^)]+\))?\s*)([^\s:]+)/',
+				function ( $matches ) use ( $variation_class ) {
+					return $matches[1] . $matches[2] . $variation_class;
+				},
```
this function can be skipped entirely, no?

 - `preg_replace( $pattern, "\\1\\2{$variation_class}", $part, $limit )`
 - `preg_replace( $pattern, "\$1\$2{$variation_class}", $part, $limit )`
